File: greed-python/greed.py
from typing import List

from outcomes import Outcome


class Greed:
    def __init__(self, outcomes_in_order: List[Outcome]):
        self.outcomes_in_order = outcomes_in_order

    def score(self, dice: List[int]) -> int:
        if len(dice) > 6:
            raise ValueError('Too many arguments')

        dice = sorted(dice)

        for outcome in self.outcomes_in_order:
            if outcome.check(dice):
                return outcome.score(dice)

        return 0

File: greed-kotlin/src/main/kotlin/nl/mvdb/greed/Greed.kt
package nl.mvdb.greed

class Greed () {
    private val outcomesInOrder = listOf(
            StraightOutcome(),
            ThreePairsOutcome(),
            NOfAKindOutcome(6),
            NOfAKindOutcome(5),
            NOfAKindOutcome(4),
            TripleOutcome(6),
            TripleOutcome(5),
            TripleOutcome(4),
            TripleOutcome(3),
            TripleOutcome(2),
            TripleOutcome(1),
            SingleOutcome(5),
            SingleOutcome(1)
        )

    fun score(dice: List<Int>): Int {
        if(dice.size > 6)
            throw IllegalArgumentException("Max 6 dice allowed")

        val sorted_dice = dice.sorted()

        for(outcome in outcomesInOrder) {
            if (outcome.check(sorted_dice))
                return outcome.score(sorted_dice)
        }

        return 0;
    }
}